### PR TITLE
Nuke war spawn rework

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request
basically
xenos based on marine count + silo count*3 - alive xenos - larva on cooldown
dying as xeno will set your larva to cooldown for 5 minutes to keep a punishment for dying

todo what to do with MODE_SILO_RESPAWN after TM
## Why It's Good For The Game
See https://hackmd.io/uhbgqI15R6Cnio47EriNZw?both
https://github.com/tgstation/TerraGov-Marine-Corps/pull/13615

Idea is that it leads to less "1 xeno burned all our slots dying 5 times as young runner" while also resolving the issue of sieges not having enough bodies by granting straight slots from silos.

## Changelog
:cl:
balance: Xeno respawn has been reworked for nuclear war. Instead of generating over time, larva slots are now determined by: xenos based on marine count + silo count*3 - alive xenos - larva on cooldown. Dying as xeno will set your larva to "cooldown" and it will be on "cooldown" for five minutes.
/:cl:

### CODE REVIEW

**Comment by ivanmixo** [comment]
Silo should def be required instead of just being an optional bonus

**Comment by TheWiznard** [comment]
am I not understanding it right or are you saying that now when you die to respawn as a new xeno no matter how big the hive is it will be 5 minutes before you can rejoin the larva queue? 

so marine dies, can be revived for like 15 minutes normally revived in maybe 2 minutes vs 5 minutes flat each time you die? 😕  
I understand the dude running like 4 larva down the drain but 5 minutes for each death? just doesn't seem right to me, couldn't we just check how many times you've died in say 15 minute intervals? like you can't respawn more than twice in 15-30 minutes

and does the larva queue also affect possessing minions? prob not but just asking

**Comment by TooFewSecrets** [comment]
> am I not understanding it right or are you saying that now when you die to respawn as a new xeno no matter how big the hive is it will be 5 minutes before you can rejoin the larva queue?
> 
> so marine dies, can be revived for like 15 minutes normally revived in maybe 2 minutes vs 5 minutes flat each time you die? 😕 I understand the dude running like 4 larva down the drain but 5 minutes for each death? just doesn't seem right to me, couldn't we just check how many times you've died in say 15 minute intervals? like you can't respawn more than twice in 15-30 minutes
> 
> and does the larva queue also affect possessing minions? prob not but just asking

As I understand it, with this PR: Xenos have a certain number of "lives" across the entire xeno faction. Spawning in as a xeno uses a "life", and it's refunded 5 minutes after that xeno dies. Judging by the PR submission text this does not actually restrict how fast the player can respawn. Lives also don't seem to "queue", meaning that a full xeno wipe minus one shrike will be back up to the exact same population (i.e. 10) within 5 minutes instead of going 2, 3, 4, etc.
I am going to assume that my interpretation is correct, and that this is merged alongside #13615 for my feedback, because the PR specifically cites it and both are part of the same design document also linked.

I personally think this PR is a pretty bad idea. A permanent flood of runners, defenders, sentinels, etc. does not sound fun for marines at all. Neither does low-tier kills having almost no inherent value. If CAS _gibs_ three idiot T1s resting on top of a signal flare that only helps marines if they can capitalize on it within 5 minutes. Or if xenos are understrength with 2 burrowed sitting around like usual, basically accomplishes nothing. If they can't... CAS just wasted a munition accomplishing nothing. Req is the only serious punishment for dying at t1 at this point, and even t2 or t3 is quickly replaced with maturity removal by xenos already sitting a tier below with a full evo bar - which entirely negates the only xeno nerf of #13615 incidentally. So really you could gib three T3s with a Keeper and still accomplish nothing because you've effectively only killed three T1s who will all respawn in 5 minutes. Compared to right now where that would be a hilariously gigantic misplay that causes the xeno team to crumple, even without marines getting any req points from the gibbed corpses.
Moreover I'm not sure how realistic it is for xenos to ever be wiped out (marine major) in this ruleset, because one living xeno can respawn the entire team every 5 minutes with call burrowed, and they very quickly become a horde of ancient t1s with #13615 . Even if it's unrealistic for xenos to come back from that, marines sitting around doing disks for half an hour instead of just winning when victory is inevitable (presumably at this point they have a shitload of req weapons and b18s vs a bunch of t1s, have fun with that) kills enjoyment and server pop for both sides.
Unless there's a gigantic auxilary nerf to xeno capabilities planned with the multi-PR rework, which I haven't seen any evidence of yet (already covered how maturity removal is realistically just a buff outside of total wipe scenarios; tactical and strategic psy points, another core part of this design doc, is yet another buff by letting turrets be used much more often; etc) this is just going to make playing xeno way easier by removing any punishment for death outside of req points.

Also, the PR description itself is terrible and basically nobody who cannot read the code understands what it means. Even I'm not entirely sure.

**Comment by Barnet2** [comment]
Numbers are up to change, but larva genning while silo is dead is not intended. That will be changed.
If the description is lacking, I'll try to give an answer:

There will be X number of larva slots based on the calculation shown in the description, and if one of those slots is taken it is removed from the number of burrowed. If the person that occupies that slot dies, it will need to regenerate for Y minutes. Once down regenerated, the slot is opened and another burrowed is added (or the person first in line for queue gets it). Each slot **should** be independent from each other in timing, but yes if there is a Xeno wipe they will come back in Y minutes. I am talking with Tivi to put it higher, at something around 8 to 10 minutes, while possibly increasing the amount of minions so you have something to do in queue (hopefully we will get the ability to queue while playing a minion soon, it is planned.) Overall, it will take a xeno player longer to come back over dying, but be at full strength when they do. Having to wait to mature to fight as a T1 sucks, hence we add the ability to play minions in queue, which can actually be quite robust.

**Comment by ivanmixo** [comment]
This also disables "hostile near silo" alert

**Comment by Aporhtonoma** [comment]
wait if benos are directly bound by the marine count does that mean the more marines they kill the less total larvas they will have (thus punishing them for doing good)

**Comment by imsxz** [comment]
> wait if benos are directly bound by the marine count does that mean the more marines they kill the less total larvas they will have (thus punishing them for doing good)

This is effectively how it works at the moment, if you wipe a significant amount of marines your larva regen will tank(albeit with a very slight bit of bonus larva from draining the body)

> Neither does low-tier kills having almost no inherent value. If CAS gibs three idiot T1s resting on top of a signal flare that only helps marines if they can capitalize on it within 5 minutes. 

this is exactly how it feels to play xeno vs marines on highpop or basically any pop when marines come in faster than you can cut them down.

**Comment by TooFewSecrets** [comment]
> this is exactly how it feels to play xeno vs marines on highpop or basically any pop when marines come in faster than you can cut them down.

If you manage to kill nine marines (the true equivalent) in a short window as xeno the result is usually a FOB retreat, which xenos are _meant_ to eventually win on average. If those people are SLs or SGs or have req stuff and they perma, their order buffs or guns or whatever aren't instantly replaced by a corpsman evolving to replace them, they're gone for good. And if marines don't retreat at that point they are normally so understrength that you can keep engaging and wipe them. This is entirely ignoring everything you get from actually cocooning 9 marines. So no, I think xenos currently are rewarded far, far more for getting marine multikills than marines will be for getting xeno multikills with this PR.

**Comment by AlexCrimson** [comment]
Would it be a terrible idea to put a timer on individual players also? To stop them sucking up new larva and feeding. Even if its just something like say, 1 minute, 5 minutes, 10 minutes etc for each subsequent death? The respawn timer could even just be removed for players with over a certain amount of xeno playtime. Like 100+ hours.

**Comment by imsxz** [comment]
> > this is exactly how it feels to play xeno vs marines on highpop or basically any pop when marines come in faster than you can cut them down.
> 
> If you manage to kill nine marines (the true equivalent) in a short window as xeno the result is usually a FOB retreat, which xenos are _meant_ to eventually win on average. If those people are SLs or SGs or have req stuff and they perma, their order buffs or guns or whatever aren't instantly replaced by a corpsman evolving to replace them, they're gone for good. And if marines don't retreat at that point they are normally so understrength that you can keep engaging and wipe them. This is entirely ignoring everything you get from actually cocooning 9 marines. So no, I think xenos currently are rewarded far, far more for getting marine multikills than marines will be for getting xeno multikills with this PR.

I mean any group of marines dying near frontline will basically always be recovered if the marines choose to recover them unless the marines are decidedly losing. I can easily kill dozens of individual suicidal marines over the course of a losing xeno round and it means nothing besides +30 psy points. Even 5 or 6 marines dying at the same time on highpop can be negligible. 

Also I don’t think it’s fair to compare 9 marines wiping to 3 t1’s.

**Comment by TooFewSecrets** [comment]
> Also I don’t think it’s fair to compare 9 marines wiping to 3 t1’s.

A large point of my original post is that with the full xeno rework in as planned in the design doc, minus req points, 3 t3s dying is essentially the same as 3 t1s dying. Xenos will still be able to "bank" evo progress while the t3 slots are full and can instantly evolve to replace them, and they will also be able to swap to other castes of the same tier for free every 15 minutes. So if a warlock dies it can instantaneously be replaced by a spitter, and that spitter can instantly be replaced by a runner. And maturity is also removed so they are all still ancient. The problem is that the original t3's player is more likely to have been good at that caste, but... literal skill issue at that point.
This isn't totally 1:1 because you do lose high tier slots with lower overall xeno population, but still, getting an ancient warlock back in 5 minutes when currently that takes like 10 even if you already have an ancient warrior/wraith and closer to 20 if you don't is a gigantic buff. And for mission-critical castes it'll be 0 minutes to get an ancient back assuming good xeno coordination around slot usages. If you lose your only boiler during a FOB siege and you have no spitters, a carrier can evo to widow then caste-swap to boiler instantly. Currently bare-minimum you'd need to have a T3 revert and sit through the whole evo timer again, and I'm pretty sure that boiler would still only be elder.
It's really going to make losses sort of not matter for xenos. Especially when marines can't recover the corpses for req points, the only thing you'd be getting out of a kill is a random t1 not existing for five minutes. If you get several kills you might lower the amount of t3s by _one_. For five minutes. Until you get to the point of the entire hive being wiped out, which... who cares at that point? And keep in mind, again, all xenos are ancient, so killing a mature/elder t3 when they would've lived if they were ancient is not going to happen anymore.

**Comment by Pariah919** [comment]
PR could use a test run, but it'll most definitely likely make xeno attrition incredibly oppressive with the maturity change
This and #13615 removes xeno attrition entirely and makes any high tension stalemate highly favor xenos if you can't outscale with req gear, generally I get the frustration of losing because some yung runy who has sub 40 hours died twice and took the slot the Pit Boss ™️would take, but this might be too far as you genuinely can never get ahead as a marine unless you utterly break the morale of the enemy team, 5 minutes isn't really that long in a game that can have rounds last for hours, if you want my opinion it should be a minimum of 15 to make wipes really matter if you're gonna include this with maturity changes.

Xenos already have most of the map under their control and their high risk gameplay in nuke war is meant to balance out their general defenders advantage and high mobility, I don't think this will work in nuclear war. It should be its own game mode, but a test could easily prove me wrong I suppose.

**Comment by AlexCrimson** [comment]
Should the stalemate not favour Xenos? Marines are supposed to be pushing objectives, and failure to do so is meant to be a loss for them, no? They are barely punished for dying, because its fairly easy to recover a dead marine and revive them. You basically need to wipe a large group of Marines to keep them down. Otherwise the attrition heavily favours the massive, nigh-unstoppable wave of Marines that will push together to one objective after another, and face little/no resistance unless there are experienced players playing strong T3/4 castes. Usually something with gas.

Personally i do not care which sides wins, nor if the text says "minor" or "major". It would just be nice to have decently long and fun rounds rather than one-sided stomps, from either side. I do not find it fun when Marines march straight to silo and do not suffer for it. Or when Xenos wipe the first group of Marines and then its a boring FoB siege for the next 1-2 hours.

I have no idea if this PR will change any of that, but Marine seems like they have a much higher WR right now, so its worth testing.

**Comment by Barnet2** [comment]
> This and #13615 removes xeno attrition entirely and makes any high tension stalemate highly favor xenos if you can't outscale with req gear, generally I get the frustration of losing because some yung runy who has sub 40 hours died twice and took the slot the Pit Boss ™️would take, but this might be too far as you genuinely can never get ahead as a marine unless you utterly break the morale of the enemy team, 5 minutes isn't really that long in a game that can have rounds last for hours, if you want my opinion it should be a minimum of 15 to make wipes really matter if you're gonna include this with maturity changes.
> 
> Xenos already have most of the map under their control and their high risk gameplay in nuke war is meant to balance out their general defenders advantage and high mobility, I don't think this will work in nuclear war. It should be its own game mode, but a test could easily prove me wrong I suppose.
> 
> tl;dr:PR could use a test run, but it'll most definitely likely make xeno attrition incredibly oppressive with the maturity change

Again, time is up to change, I agree that 5 minutes in definitely too little.

**Comment by TooFewSecrets** [comment]
> They are barely punished for dying, because its fairly easy to recover a dead marine and revive them.

This is a reward for pushing and a penalty for not pushing. If you retreat you are going to have a lot of marines going perma. Xeno reward for holding the line is conversely generating more larva for theoretically infinite respawns, and corrupted generators giving psi for more structures, primos, and eventually just acid turret spam to make pushing impossible given enough time. Marines are already rewarded for pushing, xenos are already rewarded for holding ground. You don't need to make death meaningless for xenos to accomplish that.

**Comment by AlexCrimson** [comment]
>This is a reward for pushing and a penalty for not pushing. If you retreat you are going to have a lot of marines going perma. Xeno reward for holding the line is conversely generating more larva for theoretically infinite respawns, and corrupted generators giving psi for more structures, primos, and eventually just acid turret spam to make pushing impossible given enough time. Marines are already rewarded for pushing, xenos are already rewarded for holding ground. You don't need to make death meaningless for xenos to accomplish that.

But as a round goes on, Marines scale more with req gear than Xenos do with psy points. Xenos need to wipe a push to stop it, but wiping requires killing essentially the entire Marine population, whilst they are covering each other with aim mode. The death of a single Xeno can set the hive back. Especially if its a T3 that took 10 minutes to evolve. Not only with the death, but also the req points Marines get from it.

Theoretical infinite respawns is entirely just that. In reality, Xenos are relatively squishy for how impactful their deaths are. Marines are encouraged to rambo into enemy lines and can quite literally set themselves on fire, blow themselves up, lose limbs and suffer all kinds of toxin buildup, only to get healed up fast and be back on the frontline. They can even avoid losing their weapons by using harnesses.

One of these roles is considered to be a lot of fun. The other, players claim makes them frustrated and something they cannot do for too long before switching to the other role. Can you guess which is which?

File: code/datums/gamemodes/nuclear_war.dm [hunk 1]
```diff
@@ -127,3 +136,40 @@
...
+
+	var/silo_bonus = length(possible_silos) <= 1 ? 0 : (length(possible_silos) * 3)
+	var/total_slots = get_total_joblarvaworth() / xeno_job.job_points_needed
+	var/larva_surplus = total_slots + silo_bonus - num_xenos - cooling_larvas
+	if(larva_surplus < 1)
+		return //Things are balanced, no burrowed needed
+	xeno_job.add_job_positions(FLOOR(larva_surplus, 1))
+	xeno_hive.update_tier_limits()
+
+/datum/game_mode/infestation/nuclear_war/proc/on_xeno_death()
```
Comment (ivanmixo) [inline code/datums/gamemodes/nuclear_war.dm:169]: Missing docs here but seems fine for a tm

**Review by ivanmixo — APPROVED** [review]

**Review by ivanmixo — CHANGES_REQUESTED** [review]
Look into https://github.com/tgstation/TerraGov-Marine-Corps/issues/13918
